runners/core-java/src/main/java/org/apache/beam/runners/core/TimerInternals.java

```diff
@@ -231,6 +268,56 @@ private TimerDataCoder(Coder<? extends BoundedWindow> windowCoder) {
       this.windowCoder = windowCoder;
     }
 
+    @Override
+    public void encode(TimerData timer, OutputStream outStream) throws CoderException, IOException {
+      STRING_CODER.encode(timer.getTimerId(), outStream);
+      STRING_CODER.encode(timer.getTimerFamilyId(), outStream);
+      STRING_CODER.encode(timer.getNamespace().stringKey(), outStream);
+      INSTANT_CODER.encode(timer.getTimestamp(), outStream);
+      INSTANT_CODER.encode(timer.getOutputTimestamp(), outStream);
+      STRING_CODER.encode(timer.getDomain().name(), outStream);
+    }
+
+    @Override
+    public TimerData decode(InputStream inStream) throws CoderException, IOException {
+      String timerId = STRING_CODER.decode(inStream);
+      String timerFamilyId = STRING_CODER.decode(inStream);
+      StateNamespace namespace =
+          StateNamespaces.fromString(STRING_CODER.decode(inStream), windowCoder);
+      Instant timestamp = INSTANT_CODER.decode(inStream);
+      Instant outputTimestamp = INSTANT_CODER.decode(inStream);
+      TimeDomain domain = TimeDomain.valueOf(STRING_CODER.decode(inStream));
+      return TimerData.of(timerId, timerFamilyId, namespace, timestamp, outputTimestamp, domain);
+    }
+
+    @Override
+    public List<? extends Coder<?>> getCoderArguments() {
+      return Arrays.asList(windowCoder);
+    }
+
+    @Override
+    public void verifyDeterministic() throws NonDeterministicException {
+      verifyDeterministic(this, "window coder must be deterministic", windowCoder);
+    }
+  }
+
+  /**
+   * A {@link Coder} for {@link TimerData}. To make it encoding and decoding backward compatible for
+   * DataFlow
+   */
+  class OldTimerDataCoder extends StructuredCoder<TimerData> {
```

Comment:
On second thought, I think it's safer to to keep this named TimerDataCoder and name the new coder TimerDataCoderV2. This is because I think the coder might use the class name to determine equality.

Reply:
@reuvenlax , yeah that would be great. I am updating the names.

Reply:
Done